File: linked_list/c/template_LL_queue_hj (2).c
#include <stdio.h>
#include <stdlib.h>

#define MAX_CAPACITY 3

// Node 구조체 정의 (Stack을 포함)
typedef struct Node {
    int stack[MAX_CAPACITY];  // 스택을 배열로 구현 (최대 3개까지)
    int top;                  // 스택의 최상단 인덱스
    struct Node* next;        // 다음 노드를 가리키는 포인터
} Node;

// Linked List of Stack 구조체 정의
typedef struct {
    Node* head;  // 리스트의 첫 번째 노드를 가리키는 포인터
} LinkedListOfStacks;

// 새로운 노드 생성
Node* createNode() {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->top = -1;  // 스택이 비어있음을 나타냄
    newNode->next = NULL;
    return newNode;
}

// Linked List of Stack 초기화
LinkedListOfStacks* createLinkedListOfStacks() {
    LinkedListOfStacks* list = (LinkedListOfStacks*)malloc(sizeof(LinkedListOfStacks));
    list->head = createNode();  // 빈 노드로 시작
    return list;
}

// 스택이 가득 찼는지 확인
int isFull(Node* node) {
    return node->top == MAX_CAPACITY - 1;
}

// 스택이 비었는지 확인
int isEmpty(Node* node) {
    return node->top == -1;
}

// 스택에 원소 추가
void push(LinkedListOfStacks* list, int data) {
    Node* lastNode = list->head;
    // 마지막 노드 찾기
    while (lastNode->next != NULL) {
        lastNode = lastNode->next;
    }

    if (isFull(lastNode)) {
        // 스택이 가득 찬 경우 새 노드를 추가
        Node* newNode = createNode();
        newNode->stack[++(newNode->top)] = data;  // 새 노드의 스택에 데이터 추가
        lastNode->next = newNode;  // 리스트에 새 노드 연결
    } else {
        // 스택에 데이터 추가
        lastNode->stack[++(lastNode->top)] = data;
    }
}

// 스택에서 원소 제거 및 반환
int pop(LinkedListOfStacks* list) {
    Node* current = list->head;
    Node* prev = NULL;

    // 마지막 노드 찾기
    while (current->next != NULL) {
        prev = current;
        current = current->next;
    }

    if (isEmpty(current)) {
        // 스택이 비어 있으면 이전 노드로 이동하여 삭제
        if (prev != NULL) {
            free(current);
            prev->next = NULL;
            current = prev;
        } else {
            printf("All stacks are empty!\n");
            return -1;
        }
    }

    // 현재 스택에서 원소 제거 및 반환
    return current->stack[(current->top)--];
}

// 스택 출력 (디버깅용)
void printStacks(LinkedListOfStacks* list) {
    Node* current = list->head;
    int nodeIndex = 1;
    while (current != NULL) {
        printf("Stack %d: ", nodeIndex++);
        for (int i = 0; i <= current->top; i++) {
            printf("%d ", current->stack[i]);
        }
        printf("\n");
        current = current->next;
    }
}

int main() {
    LinkedListOfStacks* stackList = createLinkedListOfStacks();

    // Push 예시
    for (int i = 1; i <= 10; i++) {
        push(stackList, i);
    }
    printStacks(stackList);

    // Pop 예시
    for (int i = 0; i < 5; i++) {
        printf("Popped: %d\n", pop(stackList));
    }
    printStacks(stackList);

    return 0;
}

File: linked_list/python/template_LL_queue_hj (2).py
class Node:
    MAX_CAPACITY = 3

    def __init__(self):
        self.stack = []
        self.next = None

    def is_full(self):
        return len(self.stack) == self.MAX_CAPACITY

    def is_empty(self):
        return len(self.stack) == 0

    def push(self, data):
        if not self.is_full():
            self.stack.append(data)
        else:
            raise Exception("Stack is full")

    def pop(self):
        if not self.is_empty():
            return self.stack.pop()
        else:
            raise Exception("Stack is empty")


class LinkedListOfStacks:
    def __init__(self):
        self.head = Node()

    def get_last_node(self):
        current = self.head
        while current.next is not None:
            current = current.next
        return current

    def push(self, data):
        last_node = self.get_last_node()
        if last_node.is_full():
            new_node = Node()
            new_node.push(data)
            last_node.next = new_node
        else:
            last_node.push(data)

    def pop(self):
        current = self.head
        prev = None

        # 마지막 노드까지 이동
        while current.next is not None:
            prev = current
            current = current.next

        if current.is_empty() and prev is not None:
            prev.next = None
            return prev.pop()
        elif not current.is_empty():
            return current.pop()
        else:
            raise Exception("All stacks are empty!")

    def print_stacks(self):
        current = self.head
        index = 1
        while current:
            print(f"Stack {index}: {current.stack}")
            current = current.next
            index += 1


# 사용 예시
stack_list = LinkedListOfStacks()

# Push 예시
for i in range(1, 11):
    stack_list.push(i)
stack_list.print_stacks()

# Pop 예시
for _ in range(5):
    print(f"Popped: {stack_list.pop()}")
stack_list.print_stacks()
